### PR TITLE
User Search Feature

### PR DESCRIPTION
**Pull request by HerrLevin**
A simple user search feature done for university. Please don't delete the branch afterwards.

### CODE REVIEW

File: .idea/dataSources.xml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (MrKrisKrisu) [inline .idea/dataSources.xml]: This idea file looks like it should be listed in .gitignore.

File: api-swagger-v0.yml [hunk 2]
```diff
@@ -882,6 +882,66 @@ paths:
...
+                  current_page:
+                    type: integer
+                    example: 1
+                  data:
+                    type: array
+                    items:
+                      $ref: '#/components/schemas/User'
+                  first_page_url:
+                    type: string
+                    example: "https://api.traewelling.de/v0/user/search/jdo?page=1"
```
Comment (MrKrisKrisu) [inline api-swagger-v0.yml]: its traewelling.de/api, not api.traewelling.de

File: api-swagger-v0.yml [hunk 3]
```diff
@@ -882,6 +882,66 @@ paths:
...
+                  from:
+                    type: integer
+                    example: 1
+                    description: Frist element on page
+                  next_page_url:
+                    type: string
+                    example: "https://api.traewelling.de/v0/user/search/jdo?page=2"
```
Comment (MrKrisKrisu) [inline api-swagger-v0.yml]: its traewelling.de/api, not api.traewelling.de

File: api-swagger-v0.yml [hunk 4]
```diff
@@ -882,6 +882,66 @@ paths:
...
+                    nullable: true
+                  path:
+                    type: string
+                    example: "https://api.traewelling.de/v0/user/search/jdo"
```
Comment (MrKrisKrisu) [inline api-swagger-v0.yml]: its traewelling.de/api, not api.traewelling.de

File: api-swagger-v0.yml [hunk 5]
```diff
@@ -882,6 +882,66 @@ paths:
...
+      security:
+        - bearerAuth: []
+      tags: [User profile]
+      summary: Searches for a user
+      parameters:
+        - in: path
+          name: searchQuery
+          schema:
+            type: string
+            example: 'jdo'
```
Comment (MrKrisKrisu) [inline api-swagger-v0.yml:896]: I think "jdo" is a bad example. Maybe use the full "john.doe"?

File: api-swagger-v0.yml [hunk 6]
```diff
@@ -1328,15 +1388,21 @@ components:
           type: string
           example: "jdoe"
         train_distance:
-          type: number
+          type: string
```
Comment (MrKrisKrisu) [inline api-swagger-v0.yml:1391]: Why was that changed to "string"?

File: api-swagger-v0.yml [hunk 7]
```diff
@@ -1328,15 +1388,21 @@ components:
...
           format: float
-          example: 454.59
+          example: "454.59"
         train_duration:
-          type: integer
-          example: 317
+          type: string
```
Comment (MrKrisKrisu) [inline api-swagger-v0.yml:1395]: ``type: string`` and ``format: integer`` looks weird for me.

File: api-swagger-v0.yml [hunk 8]
```diff
@@ -1328,15 +1388,21 @@ components:
...
+          format: integer
+          example: "317"
         points:
-          type: integer
-          example: 66
+          type: string
```
Comment (MrKrisKrisu) [inline api-swagger-v0.yml:1399]: ``type: string`` and ``format: integer`` looks weird for me.

File: app/Http/Controllers/FrontendUserController.php [hunk 9]
```diff
@@ -3,13 +3,15 @@
 namespace App\Http\Controllers;
 
 use App\Models\User;
+use Doctrine\DBAL\Exception;
```
Comment (MrKrisKrisu) [inline app/Http/Controllers/FrontendUserController.php]: Where is this DBAL\Exception used? It looks unused for me.

File: app/Http/Controllers/FrontendUserController.php [hunk 10]
```diff
@@ -96,4 +98,18 @@ public function updateProfilePicture(Request $request)
...
+        try {
+            $userSearchResponse = UserBackend::searchUser($request['searchQuery']);
+            $userSearchResponse->appends(['searchQuery' => $request['searchQuery']]);
+        } catch (HttpException $exception) {
+            return redirect()->back();
+        }
+
+        return view("search", [
+            'userSearchResponse' => $userSearchResponse,
+            'currentUser'        => Auth::user()
```
Comment (MrKrisKrisu) [inline app/Http/Controllers/FrontendUserController.php]: I think it's unnecessary to pass it to view. You can just use ``auth()->user()`` in blade.

File: app/Http/Controllers/FrontendUserController.php [hunk 11]
```diff
@@ -96,4 +98,18 @@ public function updateProfilePicture(Request $request)
         $profilePictureResponse = UserBackend::updateProfilePicture($avatar);
         return response()->json($profilePictureResponse);
     }
+
+    public function searchUser(Request $request) {
+        try {
+            $userSearchResponse = UserBackend::searchUser($request['searchQuery']);
+            $userSearchResponse->appends(['searchQuery' => $request['searchQuery']]);
```
Comment (MrKrisKrisu) [inline app/Http/Controllers/FrontendUserController.php]: I recommend to pass the variable searchQuery explicit to the view below instead of adding it to this collection. (or is this for something useful?)

File: resources/views/search.blade.php [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+@section('content')
+    <div class="container">
+        <div class="row justify-content-center">
+            <div class="col-md-8">
+                @if($userSearchResponse->count() == 0)
+                    <div class="col-md-8 offset-md-2">
+                        <div class="alert my-3 alert-danger" role="alert">
+                            Keinen Benutzer gefunden!
+                            <button type="button" class="close" data-dismiss="alert" aria-label="Close">
```
Comment (MrKrisKrisu) [inline resources/views/search.blade.php]: Is this button necessary?

File: resources/views/search.blade.php [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+                                <span aria-hidden="true">×</span>
+                            </button>
+                        </div>
+                    </div>
+                @endif
+                @foreach($userSearchResponse as $user)
+                <div class="card status mt-3" id="status-1" data-body="" data-date="Mittwoch, 07. Oktober 2020">
```
Comment (MrKrisKrisu) [inline resources/views/search.blade.php]: You should check whether the ``data-body`` and ``data-date`` are correct.

File: resources/views/search.blade.php [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+
+                            <div class="col pl-0">
+                                <a href="{{ route('account.show', ['username' => $user->username]) }}">
+                                    <h4>{{ $user->name }} <small>{{ $user->username }}</small></h4>
+                                </a>
+                                <h6>
+                                    <small>
+                                        <span class="font-weight-bold"><i class="fa fa-route d-inline"></i>&nbsp;{{ $user->train_distance }}</span><span class="small font-weight-lighter">km</span>
+                                        <span class="font-weight-bold pl-sm-2"><i class="fa fa-stopwatch d-inline"></i>&nbsp;{!! durationToSpan(secondsToDuration($user->train_duration * 60)) !!}</span>
+                                        <span class="font-weight-bold pl-sm-2"><i class="fa fa-dice-d20 d-inline"></i>&nbsp;{{ $user->points }}</span><span class="small font-weight-lighter">Pkt</span>
```
Comment (MrKrisKrisu) [inline resources/views/search.blade.php]: Is "Pkt" the correct shortcut in all languages? (Maybe use language files)

File: resources/views/search.blade.php [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+                                    <script>
+                                        window.translFollow = "{{__('profile.follow')}}";
+                                        window.translUnfollow = "{{__('profile.unfollow')}}";
+                                    </script>
+                                @else
+                                    <a href="{{ route('settings') }}" class="btn btn-sm btn-primary">{{ __('profile.settings') }}</a>
+                                @endif
+                            </div>
+                        </div>
+                        <div class="progress">
```
Comment (MrKrisKrisu) [inline resources/views/search.blade.php]: Maybe this progress bar is just an copy-paste-issue?

**Review by MrKrisKrisu — CHANGES_REQUESTED** [review]

File: api-swagger-v0.yml (hunk 5 above)
Comment (HerrLevin) [reply api-swagger-v0.yml:896]: This would not work, since the user is called jdoe and the name is "J. Doe". I think it'll surve it's purpose.

File: app/Http/Controllers/FrontendUserController.php (hunk 11 above)
Comment (HerrLevin) [reply app/Http/Controllers/FrontendUserController.php]: This is needed for the pagination. Otherwise the links throw away the search parameters, which will result in a bad request. It's the recommended practice from the docs: [https://laravel.com/docs/8.x/pagination#appending-to-pagination-links](https://laravel.com/docs/8.x/pagination#appending-to-pagination-links)

File: api-swagger-v0.yml (hunk 6 above)
Comment (HerrLevin) [reply api-swagger-v0.yml:1391]: Because it is a string in the json response.

File: api-swagger-v0.yml (hunk 8 above)
Comment (HerrLevin) [reply api-swagger-v0.yml:1399]: I can't change the openAPI standard. Sorry. :D

File: resources/views/search.blade.php (hunk 12 above)
Comment (HerrLevin) [reply resources/views/search.blade.php]: nope. Will remove it.

File: resources/views/search.blade.php (hunk 13 above)
Comment (HerrLevin) [reply resources/views/search.blade.php]: it's not needed...

File: app/Http/Controllers/FrontendUserController.php (hunk 10 above)
Comment (HerrLevin) [reply app/Http/Controllers/FrontendUserController.php]: This actually is a problem if you're not logged in. But yes, since you have to be logged in, to search, it can be handled otherwise.

File: api-swagger-v0.yml (hunk 7 above)
Comment (HerrLevin) [reply api-swagger-v0.yml:1395]: I can't change the openAPI standard. Sorry. :D

File: resources/views/search.blade.php [hunk 16]
```diff
@@ -36,10 +32,10 @@
                                     <small>
                                         <span class="font-weight-bold"><i class="fa fa-route d-inline"></i>&nbsp;{{ $user->train_distance }}</span><span class="small font-weight-lighter">km</span>
                                         <span class="font-weight-bold pl-sm-2"><i class="fa fa-stopwatch d-inline"></i>&nbsp;{!! durationToSpan(secondsToDuration($user->train_duration * 60)) !!}</span>
-                                        <span class="font-weight-bold pl-sm-2"><i class="fa fa-dice-d20 d-inline"></i>&nbsp;{{ $user->points }}</span><span class="small font-weight-lighter">Pkt</span>
+                                        <span class="font-weight-bold pl-sm-2"><i class="fa fa-dice-d20 d-inline"></i>&nbsp;{{ $user->points }}</span><span class="small font-weight-lighter">{{__('profile.points-abbr')}}</span>
                                     </small>
                                 </h6>
-                                @if($user->id !== $currentUser->id && Auth::check())
+                                @if($user->id !== Auth::user()->id && Auth::check())
```
Comment (MrKrisKrisu) [inline resources/views/search.blade.php]: Please do an ``Auth::check()`` before accessing ``Auth:user()``, not after.

**Review by MrKrisKrisu — CHANGES_REQUESTED** [review]

**Review by MrKrisKrisu — APPROVED** [review]
